proposals/ref-struct-interfaces.md

```diff
@@ -147,61 +151,113 @@ The new rules apply to the `heap` constraint:
+
+```csharp
+// Works in net9.0 but fails in all other TF
+Func<Span<char>> func;
+```
+
+This is likely to lead to customer confusion and looking at changes in `Func<T>` in the `net9.0` source wouldn't give customers any clue as to what changed.
+
+The other issue is that very subtle changes in code can cause _spooky action at a distance_ problems. Consider the following code:
+
+```csharp
+interface I1<T>
+{
+}
+```
+
+This interface would be eligable for auto-application of `allows ref struct`. If a developer comes around later though and adds a default interface method then suddenly it would not be and it would break any consumers that had already created invocations like `I1<Span<char>>`. This is a very subtle change that would be hard to track down.
 
 ### Binary breaking change
-Adding `allow T: ref struct` to an existing API is not a source breaking change. It is purely expanding the set of allowed types for an API. Need to track down if this is a binary breaking change or not. Unclear if updating the attributes of a generic parameter constitute a binary breaking change.
+
+Adding `allows ref struct` to an existing API is not a source breaking change. It is purely expanding the set of allowed types for an API. Need to track down if this is a binary breaking change or not. Unclear if updating the attributes of a generic parameter constitute a binary breaking change.
 
 ## Considerations
 
 ### Runtime support
+
 This feature requires several pieces of support from the runtime / libraries team:
+
 - Preventing default interface methods from applying to `ref struct`
-- API in `System.Reflection.Metadata` for encoding the `gpAcceptByRefLike` value.
+- API in `System.Reflection.Metadata` for encoding the `gpAcceptByRefLike` value
 - Support for generic parameters being a `ref struct`
 
-Most of this support is likely already in place. The general `ref struct` as generic parameter support is already implemented as described [here](https://github.com/dotnet/runtime/blob/main/docs/design/features/byreflike-generics.md). It's possible the DIM implementation already account for `ref struct`. But each of these items needs to be tracked down.
+Most of this support is likely already in place. The general `ref struct` as generic parameter support is already implemented as described [here][byref-like-generics]. It's possible the DIM implementation already account for `ref struct`. But each of these items needs to be tracked down.
+
+## API versioning
+
+The `allows ref struct` anti-constraint can be safely applied to a large number of generic definitions that do not have implementations. That means most delegates, interfaces and `abstract` methods can safely apply `allows ref struct` to their parameters. These are just API definitions without implementations and hence expanding the set of allowed types is only going to result in errors if they're used as type arguments where `ref struct` are not allowed.
```

Comment:
So effectively there'd be zero downside to our marking the `T` in `Func<T>`, `IEnumerable<T>`, `IList<T>`, etc. as `allow ref struct`, yes?

Reply:
Correct. The only real downside is you effectively give up on having DIM in those interfaces in the future but that seems like an unlikely future path anyways.

Reply:
Isn't there a plan to add DIMs to IList literally this upcoming release? https://github.com/dotnet/runtime/pull/95830

Reply:
yes, hence my comment https://github.com/dotnet/csharplang/pull/7865#discussion_r1464080189

Reply:
I'm not sure I grasp why adding `IList<T> where allow T : ref struct` makes it a breaking change to add a DIM to `IList<T>`. It seems like adding a DIM will be a breaking change for ref structs implementing the interface, regardless of whether the type parameter has this constraint.

```cs
interface I<T>
{
    void M()
    {
        Console.Write(1);
    }
}

class C : I<int> { } // ok
ref struct RS : I<int> { } // error: must implement I<int>.M()
```

Reply:
That's a good point. There are essentially two facets to how DIM can create breaking changes for `ref struct`. 

1. The problem where a `ref struct` implements the interface and it leads to runtime behavior because `ref stuct` can't inherit the DIM. This doesn't require any use of `allows ref struct`
2. When a DIM uses `T` in a way that is incompatible with `allows ref struct`. This is similar to the future proofing concerns around `abstract` methods. 

Overall (2) feels more like a theoretical concern vs. a practical one. A DIM that uses `T` in `allows ref struct` would be hard pressed to create a lifetime issue because all such `T` are going to be _safe-to-return_. The `this` is _safe-to-return_ hence all values sourced from it are going to be too. 

The only actual concern I can think of is for (2) you fashion some API that takes and returns `ref` values of type `T` **and** there is a `ref struct` that implements the interface and sources `T` from a field. 

```csharp
interface I1<T> where T : allows ref struct
{
  ref T M(ref T p);
}

ref struct S : I1<Span<char> {
  Span<char> span;
  public Span<char> M(ref Span<char> span) { ... } 
}
```

If you toyed around with the lifetimes there a bit I'm sure you could find a way  to make it difficult to implement on the `ref struct`. Feels very corner case though.

Reply:
> It seems like adding a DIM will be a breaking change for ref structs implementing the interface

BTW, I wanted to kind of.. re-emphasize this. DIM, as far as I know, was supposed to be a feature which permitted library authors to evolve interfaces without imposing breaking changes on their consumers. It feels like this feature actually takes that away. Library authors can no longer rely on it. We should be very deliberate in deciding whether that is acceptable, or, otherwise, we should consider devising some way to declare a DIM such that it can be called on a type parameter constrained to ref struct, and library authors who are concerned about compatibility can use that.

Reply:
>  It feels like this feature actually takes that away

How does it take it away though? I've already pointed out that there are existing cases where adding a DIM member leads to runtime exceptions. This is one of the hesitations runtime has in adding DIM to existing API. So this is already very much an existing problem. All this proposal does is add one more case of it. 

>  we should consider devising some way to declare a DIM such that it can be called on a type parameter constrained to ref struct, and library authors who are concerned about compatibility can use that.

The issue is not a language one, it is a runtime one. The act of invoking a DIM forces a box. This was called out during the implementation of DIM as a potential future limitation if `ref struct` were allowed to implement interfaces. The decision was made that it was a reasonable trade off. If in the future the runtime wants to go back on that then we could evaluate the intersection of DIM and `ref struct`. 

I think that is unlikely though. As mentioned there are existing limitations of DIM and those are more prevalent than the `ref struct` one. Given that we haven't felt the need to revisit those limitations I think it's unlikely runtime will revisit this one.

Reply:
> I've already pointed out that there are existing cases where adding a DIM member leads to runtime exceptions.

I wasn't able to find the description of these existing cases, could you please point me to it?

Reply:
https://github.com/dotnet/csharplang/blob/main/meetings/2018/LDM-2018-10-17.md#diamond-inheritance

Reply:
Thanks, I think I'm satisfied that we're not entering new territory with this interaction between ref structs and DIMs, it just means my dreams were never reality 😄